### PR TITLE
New Control - ViewPicker - Fixing Issue 1439

### LINKED ISSUE #1439: Request to add ViewPicker similar to SitePicker and ListPicker
#### Category
[x] Enhancement
[ ] Bug
[ ] Question

#### Version
Please specify what version of the library you are using: [3.12.0]

#### Expected / Desired Behavior / Question
Use a controls to get the views from a list similar to SitePicker and ListPicker

#### Observed Behavior
A ViewPicker control doesn't currently exists

Thanks!

### PR DESCRIPTION
**Pull request by NishkalankBezawada**
| Q               | A
| --------------- | ---
| Bug fix?        | [ ]
| New feature?    | [x]
| New sample?      | [ ]
| Related issues?  | fixes #1439 

#### What's in this Pull Request?

Added a new control 'ViewPicker' which enables to select the list views just like ListPicker, SitePicker, FieldPicker controls

#### Properties used

| Property | Type | Required | Description |
| ---- | ---- | ---- | ---- |
| context | BaseComponentContext | yes | The context object of the SPFx loaded webpart or customizer. |
| className | string | no | If provided, additional class name to provide on the dropdown element. |
| disabled | boolean | no | Whether or not the view picker control is disabled. |
| filter | string | no | Filter views from Odata query. |
| label | string | no | Label to use for the control. |
| listId | string | no | The List Id of the list. |
| placeholder | string | no | Placeholder label to show in the dropdown. |
| orderBy | Enum | no | How to order the set of views (By ID or Title). |
| selectedView | string OR string[] | no | Keys(View Ids) of the selected item(s). If you provide this, you must maintain selection state by observing onSelectionChanged events and passing a new value in when changed. |
| multiSelect | boolean | no | Optional mode indicates if multi-choice selections is allowed. Default to `false`. |
| showBlankOption | boolean | no | Whether or not to show a blank option. Default to `false`. |
| viewsToExclude | string[] | no | Defines view titles which should be excluded from the view picker control. |
| webAbsoulteUrl | string | no | Absolute Web Url of target site (user requires permissions) |
| onSelectionChanged | (newValue: string OR string[]): void | no | Callback function when the selected option changes. |

#### Screens

Below is the implementation of the control

```TypeScript
    <ViewPicker context={this.props.context}
        listId={"9f3908cd-1e88-4ab3-ac42-08efbbd64ec9"}
        placeholder={'Select list view(s)'}
        orderBy={orderBy.Title}
        multiSelect={true}
        onSelectionChanged={this.onViewPickerChange} />
```
image

- The `onSelectionChanged` change event returns the selected view(s) and can be implemented as follows in your webpart:

```TypeScript
  private onViewPickerChange = (views: string | string[]) => {
    console.log("Views:", views);
  }
```

Below is the screen of the control from workbench

image

Below is the screen of the control in single selection mode

image

Below is the screen of the control in multi selection mode

image

Below are the sample screen of the list views. 

image

//Nishkalank Bezawada

### CODE REVIEW

File: package-lock.json
Comment (NishkalankBezawada) [inline package-lock.json:1]: Hello,

package-lock.json is not required to be merged in this commit. Kindly resolve this conflicts. 

//Nishkalank

**Comment by NishkalankBezawada** [comment]
Hello Alex @AJIXuMuK

I was a bit away with some medical stuff. I will soon update the requested changes. 

Thanks & Have a wonderful weekend,
Nishkalank

**Review by NishkalankBezawada — COMMENTED** [review]
Corrected as suggested. Kindly review. 

Thanks,
Nishkalank

**Review by NishkalankBezawada — COMMENTED** [review]
Hello @AJIXuMuK 

Corrected and completed the changes as suggested. Kindly review.

Thanks,
Nishkalank

**Comment by NishkalankBezawada** [comment]
Hello @AJIXuMuK @joelfmrodrigues 

Corrected and completed the changes as suggested. Kindly review.

Thanks,
Nishkalank

File: src/common/utilities/GeneralHelper.ts [hunk 1]
```diff
@@ -4,6 +4,7 @@ import * as strings from 'ControlStrings';
 
 export const IMG_SUPPORTED_EXTENSIONS = ".gif,.jpg,.jpeg,.bmp,.dib,.tif,.tiff,.ico,.png,.jxr,.svg";
 
+import * as _ from '@microsoft/sp-lodash-subset';
```
Comment (AJIXuMuK) [inline src/common/utilities/GeneralHelper.ts:7]: Why do you need this `import`

File: src/controls/viewPicker/IViewPicker.ts [hunk 2]
```diff
@@ -0,0 +1,100 @@
+import { BaseComponentContext } from '@microsoft/sp-component-base';
+import { ISPView } from "../../../src/common/SPEntities";
```
Comment (AJIXuMuK) [inline src/controls/viewPicker/IViewPicker.ts:2]: Please, use the same quotes style: either `'` or `"`

File: src/controls/viewPicker/ViewPicker.tsx [hunk 3]
```diff
@@ -0,0 +1,200 @@
...
+          results: options
+        });
+        
+    }
+
+    /**
+     * Set the currently selected views(s);
+     */
+    private setSelectedViews(): void {
+      this._selectedView = cloneDeep(this.props.selectedView);
```
Comment (AJIXuMuK) [inline src/controls/viewPicker/ViewPicker.tsx:125]: Why do you have `selectedView` as both field and property in the state?
I don't get the point to be honest.

File: src/controls/viewPicker/ViewPicker.tsx [hunk 4]
```diff
@@ -0,0 +1,200 @@
...
+          dropdownProps.multiSelect = true;
+          dropdownProps.selectedKeys = selectedView as string [];
+        }else {
+          dropdownProps.selectedKey = selectedView as string;
+        }
+        // Renders content
+        return (
+          <div className={styles.viewPicker}>
+            {loading && <Spinner className={styles.spinner} size={SpinnerSize.xSmall}/>}
+            <Dropdown {...dropdownProps} />
```
Comment (AJIXuMuK) [inline src/controls/viewPicker/ViewPicker.tsx:195]: is it shown during loading? if yes - why?

File: src/services/SPService.ts [hunk 5]
```diff
@@ -735,4 +736,55 @@ export default class SPService implements ISPService {
       return substringSearch ? fieldValue.indexOf(lowercasedFilterText) > -1 : fieldValue.startsWith(lowercasedFilterText);
     });
   }
+
+  /**
+   * Gets the collection of view for a selected list
+   */
+  public async getViews(listId?: string, _webAbsoluteUrl?:string, _orderBy?: orderBy, filter?: string): Promise<ISPViews> {
```
Comment (AJIXuMuK) [inline src/services/SPService.ts:743]: why prefix with `_` **some** of the parameters?
Could you please use the same standard naming approach?

**Review by AJIXuMuK — CHANGES_REQUESTED** [review]
@NishkalankBezawada - sorry for a long delay.
Added a few more comments.

File: src/common/utilities/GeneralHelper.ts (hunk 1 above)
Comment (NishkalankBezawada) [reply src/common/utilities/GeneralHelper.ts]: Hello @AJIXuMuK,

Removed this unwanted ```import``` in the latest commit

Thanks,
Nishkalank

File: src/controls/viewPicker/IViewPicker.ts (hunk 2 above)
Comment (NishkalankBezawada) [reply src/controls/viewPicker/IViewPicker.ts]: Hello @AJIXuMuK 

Updated as suggested.

Thanks,
Nishkalank

File: src/controls/viewPicker/ViewPicker.tsx (hunk 4 above)
Comment (NishkalankBezawada) [reply src/controls/viewPicker/ViewPicker.tsx:189]: Updated in the latest commit as suggested.

Thanks,
Nishkalank

File: src/controls/viewPicker/ViewPicker.tsx (hunk 3 above)
Comment (NishkalankBezawada) [reply src/controls/viewPicker/ViewPicker.tsx:120]: This global variable is used to send the selected value to the callback function ```onSelectionChanged``` as below code did not work with state. 

```
  private onChange = (event: React.FormEvent<HTMLDivElement>, option: IDropdownOption, index?: number): void => {
    const { multiSelect, onSelectionChanged } = this.props;

    if (multiSelect) {
      let selectedViews = this.state.selectedView ? cloneDeep(this.state.selectedView) as string[] : [];

      if (option.selected) {
        selectedViews.push(option.key.toString());
      }
      else {
        selectedViews = selectedViews.filter(view => view !== option.key);
      }
      this.setState({
          selectedView : selectedViews 
      })
    }
    else {
      this.setState({
          selectedView : option.key.toString()
      })
    }
    if (onSelectionChanged) {
      onSelectionChanged(cloneDeep(this._selectedView));
    }
  }
```

Hence referred this code from ListPicker control as in below screen shot
image

Thanks,
Nishkalank

File: src/services/SPService.ts (hunk 5 above)
Comment (NishkalankBezawada) [reply src/services/SPService.ts]: Updated as suggested in the latest commit. 

Thanks,
Nishkalank

**Comment by NishkalankBezawada** [comment]
Hello @AJIXuMuK @joelfmrodrigues

Corrected and completed the changes dated 24-06-2023 as suggested and added two more commits to this PR. Kindly review.

Thanks,
Nishkalank

File: src/controls/viewPicker/ViewPicker.tsx (hunk 3 above)
Comment (AJIXuMuK) [reply src/controls/viewPicker/ViewPicker.tsx:120]: @NishkalankBezawada - `ListPicker` doesn't have selected lists in the state, only as a field.
Your code will work if you just pass `selectedViews` or `option.key.toString()` to the `onSelectionChanged`.
Please, leave either field or state.

**Review by AJIXuMuK — CHANGES_REQUESTED** [review]
@NishkalankBezawada - please modify the logic with the `_selectedView`.
And also resolve the conflicts.

File: src/controls/viewPicker/ViewPicker.tsx (hunk 3 above)
Comment (NishkalankBezawada) [reply src/controls/viewPicker/ViewPicker.tsx]: Hello @AJIXuMuK 
Your suggestion worked like a charm. Thanks for your guidance. Updated accordingly in the latest commit.

Thanks,
Nishkalank

**Comment by NishkalankBezawada** [comment]
Hello @AJIXuMuK & @joelfmrodrigues 

About the merge conflicts, I need your guidance as i am not able to see those as below (note : I am targetting issue-1439-viewpicker branch)


image

Thanks,
Nishkalank

**Comment by AJIXuMuK** [comment]
@NishkalankBezawada you need to sync your forked repo with the origin. 
After that git pull
Then switch to your branch and git merge dev. 
You will see the conflicts

**Comment by NishkalankBezawada** [comment]
Thanks @AJIXuMuK, seems the forked 'dev' repo is upto date. 
image

But, this repo isn't, and i guess i have forked it from 'master' not the 'dev'
image

Please suggest.

Thanks,
Nishkalank

**Comment by NishkalankBezawada** [comment]
Thanks alot @AJIXuMuK for your guidance, I have learnt alot of new things. 

//Nishkalank

**Comment by joelfmrodrigues** [comment]
@NishkalankBezawada we merged a PR that makes the test control much simpler, but this means that your PR now has conflicts.
Could you please sync your branch and resolve the ControlsTest conflicts? In summary, the visibility of each control is now controlled by a switch that is associated with a state property, and the controls are only displayed if the user toggles the switch. Should be easy to update by just copying the test control you added and adding it again to the latest version of the ControlsTest.tsx file

Let me know if you run into issues

**Comment by NishkalankBezawada** [comment]
Hello @joaojmendes 

I have now merged the conflicts on my repo. 

Thanks,
Nishkalank

**Comment by NishkalankBezawada** [comment]
Hello @joelfmrodrigues, 

As per your suggestion, i have modified the component, and added my control to it. It works with my branch now. 

image


Thanks,
Nishkalank

**Comment by joelfmrodrigues** [comment]
@NishkalankBezawada Sorry for the very long delay that the entire process took and for all the updates you did along it. This is now merged 🙂

**Comment by NishkalankBezawada** [comment]
> @NishkalankBezawada Sorry for the very long delay that the entire process took and for all the updates you did along it. This is now merged 🙂

Thanks @joelfmrodrigues This was long awaited one ☺️.

**Pull request merged**